libsignal-protocol/tests/libsignal-protocol-c-tests.rs

```diff
@@ -400,3 +411,124 @@ fn test_optional_one_time_pre_key() {
+    let bob_store = sig::store_context(
+        &ctx,
+        InMemoryPreKeyStore::default(),
+        InMemorySignedPreKeyStore::default(),
+        InMemorySessionStore::default(),
+        InMemoryIdentityKeyStore::new(registration_id, &bob_identity_key_pair),
+    )
+    .unwrap();
+
+    let bob_local_registration_id = bob_store.registration_id().unwrap();
+    let bob_signed_pre_key_pair = sig::generate_signed_pre_key(
+        &ctx,
+        &bob_identity_key_pair,
+        22,
+        SystemTime::now(),
+    )
+    .unwrap();
+    let bob_signed_pre_key_public_serialized = bob_signed_pre_key_pair
+        .key_pair()
+        .public()
+        .serialize()
+        .unwrap();
+    let bob_signed_pre_key_signature = sig::calculate_signature(
+        &ctx,
+        &bob_identity_key_pair.private(),
+        bob_signed_pre_key_public_serialized.as_slice(),
+    )
+    .unwrap();
+    bob_store
+        .store_signed_pre_key(&bob_signed_pre_key_pair)
+        .unwrap();
+
+    // Generate pre key for bob
+    let mut bob_pre_keys = sig::generate_pre_keys(&ctx, 2, 4).unwrap();
+    let bob_pre_key = bob_pre_keys.next().unwrap();
+    bob_store.store_pre_key(&bob_pre_key).unwrap();
+
+    let bob_pre_key_bundle = PreKeyBundle::builder()
+        .registration_id(bob_local_registration_id)
+        .identity_key(&bob_identity_key_pair.public())
+        .device_id(1)
+        .pre_key(bob_pre_key.id(), &bob_pre_key.key_pair().public())
+        .signed_pre_key(
+            bob_signed_pre_key_pair.id(),
+            &bob_signed_pre_key_pair.key_pair().public(),
+        )
+        .signature(bob_signed_pre_key_signature.as_slice())
+        .build()
+        .unwrap();
+
+    // Have Alice process Bob's pre key bundle
+    alice_session_builder
+        .process_pre_key_bundle(&bob_pre_key_bundle)
+        .unwrap();
+
+    // create alice's session cipher
+    let alice_session_cipher =
+        sig::SessionCipher::new(&ctx, &alice_store, &bob_address).unwrap();
+
+    // Let Alice send a message to bob
+    let msg = "Hello bob!";
+    let outgoing_message =
+        alice_session_cipher.encrypt(msg.as_bytes()).unwrap();
+
+    // Convert to an incoming message (this is technically a downcast from
+    // CiphertextMessage to PreKeySignalMessage)
+    let incoming_message =
+        PreKeySignalMessage::try_from(outgoing_message).unwrap();
+
+    let has_pre_key_id = incoming_message.has_pre_key_id();
+    assert!(has_pre_key_id);
+
+    // Create bob's session cipher
+    let bob_session_cipher =
+        sig::SessionCipher::new(&ctx, &bob_store, &alice_address).unwrap();
+
+    // Decrypt message to bob
+    let decrypted_msg = bob_session_cipher
+        .decrypt_pre_key_message(&incoming_message)
+        .unwrap();
+    assert_eq!(msg, std::str::from_utf8(decrypted_msg.as_slice()).unwrap());
+
+    // Let bob send a message to Alice
+    let msg = "Hi Alice!";
```

Comment:
Lol, we needed about 100 lines of setup and key exchanges when the actual interesting bit (sending a message) takes less than 10 lines.

Is sending messages between clients always going to be this annoying/tedious/stateful, or will higher-level applications be able simplify the process later on?

Reply:
Once the communication between two clients have been established there is not that much code that is involved in sending and receiving messages.
Most of this code will be invoked during device setup and during messaging session setup. For example:
- At device setup identity key and pre keys will be generated and the server will generate registration id that the client stores.
- The pre key will be fetched from the server and processed once the user decides to start a communication with another user.

Reply:
That makes a lot of sense. So it seems tedious because we're doing all the setup in one place when it would normally be done once per conversation (i.e. Alice starts talking to Bob) then a bit more before sending a message.

Reply:
Exactly